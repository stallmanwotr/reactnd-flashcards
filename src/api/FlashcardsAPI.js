import { AsyncStorage } from 'react-native'

const FLASHCARD_STORAGE_KEY = 'Udaci:Flashcards:All'

/**
 * Builds the initial application state.
 */
function getInitialDecks() {
    return {};
}

/**
 * Gets all decks along with their titles, questions, and answers.
 *
 * Example:
 *
 * {
 *   React: {
 *     title: 'React',
 *     questions: [{
 *         question: 'What is React?',
 *         answer: 'A library for managing user interfaces'
 *       }, {
 *         question: 'Where do you make Ajax requests in React?',
 *         answer: 'The componentDidMount lifecycle event'
 *       }]
 *   },
 *   JavaScript: {
 *     title: 'JavaScript',
 *     questions: [{
 *         question: 'What is a closure?',
 *         answer: 'The combination of a function and ... declared.'
 *       }]
 *   }
 * }
 *
 * @return A Promise object.
 */
export function getDecks() {
    return AsyncStorage.getItem(FLASHCARD_STORAGE_KEY)
        .then(value =>
            (!value) ? getInitialDecks() : JSON.parse(value)
        );
}

/**
 * The deck associated with the id argument, otherwise undefined.
 *
 * @return A Promise object.
 */
export function getDeck(deckId) {
  return getDecks()
      .then(decks => decks[deckId]);   // undefined if not set
}

/**
 * Removes all decks from the storage.
 *
 * @return A Promise object.
 */
export function clearDecks() {
    const decks = getInitialDecks();
    return AsyncStorage.setItem(FLASHCARD_STORAGE_KEY, JSON.stringify(decks));
}


/**
 * Adds a new deck with the specified title.
 *
 * @param title The title of the associated deck.
 * @return A Promise object.
 */
export function saveDeckTitle(title) {
    return AsyncStorage.mergeItem(FLASHCARD_STORAGE_KEY, JSON.stringify({
        [title]: {
            title,
            questions: []
        }
    }));
}

/**
 * Adds a new card to the indicated deck.
 *
 * @param title The title of the associated deck.
 * @param card The card to be added to the list of questions.
 * @return A Promise object.
 */
export function addCardToDeck(title, card) {
    return getDecks()
        .then(decks => {
            const deck = decks[title];
            if (deck && deck.questions) {
                deck.questions.push(card);
                return deck.questions;
            }
            return [];
        })
        .then(questions => {
            return AsyncStorage.mergeItem(FLASHCARD_STORAGE_KEY, JSON.stringify({
                [title]: {
                    questions
                }
            }));
        });
}

