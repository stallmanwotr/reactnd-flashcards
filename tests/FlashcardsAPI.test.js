import React from 'react';
import * as FlashcardsAPI from '../src/api/FlashcardsAPI.js';

//import renderer from 'react-test-renderer';

/** Simple test to try out Jest snapshots. */
test('Simple Jest test', () => {
   const decks = { hello : 'world' };
   expect(decks).toMatchSnapshot();
});

test('Add a new deck', async () => {
    const title = 'Test Title 1';

    // NOT WORKING: 'decks' is always the empty object.
    // works fine when run in non-test code on the android emulator .

    await FlashcardsAPI.clearDecks();

    await FlashcardsAPI.saveDeckTitle(title);

    const decks = await FlashcardsAPI.getDecks();

    expect(decks).toMatchSnapshot();
});

test('Add a new deck (v2)', (done) => {
    const title = 'Test Title 1';

    // NOT WORKING: 'decks' is always the empty object.
    // works fine when run in non-test code on the android emulator .

    FlashcardsAPI.clearDecks().then(() => {
        FlashcardsAPI.saveDeckTitle(title).then(() => {
            FlashcardsAPI.getDecks().then((decks) => {
                expect(decks).toMatchSnapshot();
                done();
            })
        })
    })
});

