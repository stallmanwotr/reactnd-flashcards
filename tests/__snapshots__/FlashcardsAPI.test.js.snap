// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add a new deck (v2) 1`] = `Object {}`;

exports[`Add a new deck 1`] = `Object {}`;

exports[`Simple Jest test 1`] = `
Object {
  "hello": "world",
}
`;
